api/routers/report.py

```diff
@@ -102,37 +111,107 @@ async def update_reports(old_user_id: int, new_user_id: int, token: str):
-async def insert_report(token: str, detections: List[detection]):
+async def insert_report(request : Request,
+    detections: List[detections],
+    manual_detect: int = Query(0, ge=0, le=1),
+    ):
     '''
-        Work in progress 
-        Insert report.
+        Inserts detections to the plugin database.
     '''
-    await verify_token(token, verification='verify_ban', route='[POST]/v1/report/')
+    client_host = request.client.host
+    success = JSONResponse(status_code=status.HTTP_200_OK, content=jsonable_encoder({'OK': 'OK'}))
+    bad_data = JSONResponse(status_code=status.HTTP_400_BAD_REQUEST, content=jsonable_encoder({'error': 'There was an error with your request. Please contact plugin support.'}))
+
+    # remove duplicates
+    df = pd.DataFrame([d.dict() for d in detections])
+    df.drop_duplicates(
+        subset=['reporter', 'reported', 'region_id'], inplace=True)
+
+    # data validation, there can only be one reporter, and it is unrealistic to send more then 5k reports.
+    sender = df["reporter"].values[0]
+    if len(df) > int(report_maximum) or df["reporter"].nunique() > 1:
+        logger.warning(f'IP: {client_host} | Sender {sender} | Too Many Reports or Multiple Reporters!')
+        return bad_data
+
+    # data validation, checks for correct timing
+    now = int(time.time())
+    df_time = df.ts
+    mask = (df_time > int(now + int(front_time_buffer))) | (df_time < int(now - int(back_time_buffer)))
+    if len(df_time[mask].values) > 0:
+        logger.warning(f'IP: {client_host} | Sender {sender} | Data contains out of bounds time!')
+        return bad_data
+
+    # Successful query
+    logger.debug(f"Received: {len(df)} from: {df['reporter'].unique()}")
+
+    # 1) Get a list of unqiue reported names and reporter name
+    names = list(df['reported'].unique())
+    names.extend(df['reporter'].unique())
+
+    # 1.1) Normalize and validate all names
+    clean_names = [await to_jagex_name(name) for name in names if await is_valid_rsn(name)]
+
+    # 2) Get IDs for all unique names
+    data = await sql_select_players(clean_names)
+
+    # 3) Create entries for players that do not yet exist in Players table
+    existing_names = [d["normalized_name"] for d in data]
+    new_names = set([name for name in clean_names]).difference(existing_names)
+
+    # 3.1) Insert new names and get those player IDs from step 3
+    if new_names:
+        param = [{"name": name, "normalized_name": name} for name in new_names]
+        await batch_function(sql_insert_player, param)
+        data.extend(await sql_select_players(new_names))
+
+    # 4) Insert detections into Reports table with user ids
+    # 4.1) add reported & reporter id
+    df_names = pd.DataFrame(data)
+
+    try:
+        df = df.merge(df_names, left_on="reported", right_on="name")
+    except KeyError as e:
+        logger.debug(f'IP: {client_host} | Sender {sender} | Key Error: {df} '+f'{df.columns}, {e}')
+        return bad_data
```

Comment:
bad practice, you get KeyError because the df is empty

Reply:
Will resolve